backend/creditsWorker.js

```diff
@@ -0,0 +1,190 @@
+const GRANULAR_ENTRIES_PER_COARSE_ENTRIES = 288
+
+// Timestamp when this worker instance started. Used to avoid charging while the
+// server is not running
+const startTime = Date.now()
+
+/**
+ * Updates the credit balance and history for a given billable entity.
+ * Can handle charging based on size/time or adding credits directly.
+ * Uses an event queue for atomic updates per entity.
+ *
+ * @param billableEntity - The identifier of the entity to bill.
+ * @param type - The type of update: 'charge' (based on size/time) or 'credit'
+ * (direct addition).
+ * @param amount - For 'charge', this is the current total size (bytes). For 'credit', the amount to add.
+ */
+const updateCredits = async (billableEntity: string, type: 'credit' | 'charge', amount: number) => {
+  const granularHistoryKey = `_private_ownerBalanceHistoryGranular_${billableEntity}`
+  // Use a queue to ensure atomic updates
+  await sbp('okTurtles.eventQueue/queueEvent', granularHistoryKey, async () => {
+    const now = new Date()
+    const date = now.toISOString()
+
+    // Fetch the existing history. bypassCache ensures we get the latest state before calculating.
+    // Performance concern: Reading/parsing about 1000 entries per contract
+    const granularHistoryList = await sbp('chelonia.db/get', granularHistoryKey, { bypassCache: true }) ?? '[]'
+    const granularHistory = JSON.parse(granularHistoryList)
+
+    // Trim the history if it exceeds the maximum length. Remove the oldest
+    // entry (at the end)
+    if (granularHistory.length >= GRANULAR_HISTORY_MAX_ENTRIES) {
+      granularHistory.splice(GRANULAR_HISTORY_MAX_ENTRIES)
+    }
+
+    // Get the balance from the most recent history entry.
+    const previousBalance = BigInt(granularHistory[0]?.balance || 0)
+
+    if (type === 'charge') {
+      // Get timestamp of the last recorded event. Default to 'now' if there's
+      // no history, ensuring elapsed time is calculated correctly for the first
+      // charge.
+      const previousTime = new Date(granularHistory[0]?.date).getTime() || now.getTime()
+      // Calculate time elapsed in seconds since the later of:
+      // 1. The last event time stored in history.
+      // 2. The time this worker process started (`startTime`).
+      // This prevents charging for time before the last update or before the
+      // service was running.
+      const timeElapsed = Math.max(0, now.getTime() - Math.max(previousTime, startTime))
+      // Calculate credits used: size (amount) * time * rate.
+      const creditsUsed = BigInt(Math.floor(amount * timeElapsed / 1000)) * CREDITS_PER_BYTESECOND
+
+      // Calculate new balance
+      const balance = (previousBalance - creditsUsed).toString(10)
+
+      // Prepare the new history entry
+      granularHistory.unshift({
+        type: 'charge',
+        date,
+        sizeTotal: amount,
+        credits: creditsUsed.toString(10),
+        period: `${new Date(now - timeElapsed).toISOString()}/${date}`,
```

Comment:
Shouldn't this be more like `<from> - <to>`?

Reply:
Are you talking about the date format? The ISO 8601 way of expressing a period is `{start}/{end}`. For legacy reasons, also `{start}--{end}` is acceptable, but that requires mutual agreement (i.e., being aware of the use of `--`).